simde/arm/neon/cgtz.h

```diff
@@ -0,0 +1,379 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright:
+ *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2020      Christopher Moore <moore@free.fr>
+ */
+
+#if !defined(SIMDE_ARM_NEON_CGTZ_H)
+#define SIMDE_ARM_NEON_CGTZ_H
+
+#include "cgt.h"
+#include "dup_n.h"
+#include "types.h"
+
+HEDLEY_DIAGNOSTIC_PUSH
+SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+SIMDE_BEGIN_DECLS_
+
+SIMDE_FUNCTION_ATTRIBUTES
+simde_uint32x2_t
+simde_vcgtz_f32(simde_float32x2_t a) {
+  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
+    return vcgtz_f32(a);
+  #elif defined(SIMDE_NATURAL_VECTOR_SIZE) && (SIMDE_NATURAL_VECTOR_SIZE >= 64)
```

Comment:
Why not just make the whole function

```c
  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
    return vcgtz_f32(a);
  #else
    return simde_vcgt_f32(a, simde_vdup_n_f32(SIMDE_FLOAT32_C(0.0)));
  #endif
```

?  Or maybe

```c
  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
    return vcgtz_f32(a);
  #elif defined(SIMDE_VECTOR_SUBSCRIPT_SCALAR)
    simde_float32x2_private a_ = simde_float32x2_to_private(a);
    simde_uint32x2_private r_;

    r_.values = HEDLEY_REINTERPRET_CAST(__typeof__(r_.values), a_.values > SIMDE_FLOAT32_C(0.0));

    return simde_uint32x2_from_private(r_);
  #else
    return simde_vcgt_f32(a, simde_vdup_n_f32(SIMDE_FLOAT32_C(0.0)));
  #endif
```

The only time I can think of where we wouldn't want to fall back on this implementation is on an architecture which supports comparison to zero without having to involve another register, and the compiler is smart enough to auto-vectorize the loop to that instruction.  ARM is the only architecture I'm aware of that has comparison to zero built in (though there could be others), but even then putting the vector extension version first should take care of that.

What do you think?  It would certainly simplify things, and I'm having trouble thinking of any disadvantages.

Reply:
> Why not just make the whole function
> 
> ```c
>   #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
>     return vcgtz_f32(a);
>   #else
>     return simde_vcgt_f32(a, simde_vdup_n_f32(SIMDE_FLOAT32_C(0.0)));
>   #endif
> ```
> 

I didn't want to do this because if there is no SIMD, this will create a loop for each sub-function and I am not sure whether all compilers will auto-vectorize and combine them.

> ? Or maybe
> 
> ```c
>   #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
>     return vcgtz_f32(a);
>   #elif defined(SIMDE_VECTOR_SUBSCRIPT_SCALAR)
>     simde_float32x2_private a_ = simde_float32x2_to_private(a);
>     simde_uint32x2_private r_;
> 
>     r_.values = HEDLEY_REINTERPRET_CAST(__typeof__(r_.values), a_.values > SIMDE_FLOAT32_C(0.0));
> 
>     return simde_uint32x2_from_private(r_);
>   #else
>     return simde_vcgt_f32(a, simde_vdup_n_f32(SIMDE_FLOAT32_C(0.0)));
>   #endif
> ```
> 
> The only time I can think of where we wouldn't want to fall back on this implementation is on an architecture which supports comparison to zero without having to involve another register, and the compiler is smart enough to auto-vectorize the loop to that instruction. ARM is the only architecture I'm aware of that has comparison to zero built in (though there could be others), but even then putting the vector extension version first should take care of that.
> 
> What do you think? It would certainly simplify things, and I'm having trouble thinking of any disadvantages.
>

Frankly I didn't understand what you can do in a `SIMDE_VECTOR_SUBSCRIPT_*`. But I would prefer to force my solution than to hope that the compiler vectorization works. I don't see how the compiler could realize how to decompose e.g. the vrsra intrinsics.

Reply:
> I didn't want to do this because if there is no SIMD, this will create a loop for each sub-function and I am not sure whether all compilers will auto-vectorize and combine them.

That's a valid concern; compilers aren't great about fusing loops across functions like that.  Moving the vector extension version should take care of that, at least for compilers which support GCC-style vector extensions (basically, "not MSVC").

I guess the question is whether that is a bigger concern than an unnecessary splat.  My instinct is that it would be better to optimize for platforms where we have some hope of running at a reasonable speed; if we can't vectorize this function I feel like the battle is already lost.  However, after thinking about it a bit more I'm starting to think you're right here; NEON is the only platform I'm aware of with compare to zero instructions instead of making you compare to a zeroed register…

> Frankly I didn't understand what you can do in a `SIMDE_VECTOR_SUBSCRIPT_*`. But I would prefer to force my solution than to hope that the compiler vectorization works. I don't see how the compiler could realize how to decompose e.g. the vrsra intrinsics.

GCC-style vector extensions are a bit limited, but if they are supported they are extremely reliable (probably because they're so limited).  If they are available it's safe to assume that they'll provide the best result.

> Edit: I know you don't like my implementation

It's not that I dislike it; with a couple very small exceptions I think the code is correct, I'm just trying to figure out how we could reuse other functions so we can take advantage of optimizations there.

> (originally I didn't realize that SIMDE_NATURAL_VECTOR_SIZE could be undefined),

Okay, that explains some things.  I was really confused about why you're testing to see if it is >= 64.

If all you're interested in is whether the platform supports SIMD, I think what you're looking for here is the `SIMDE_ASSUME_VECTORIZATION` macro, which has been part of SIMDe for a long time.  Now that `SIMDE_NATURAL_VECTOR_SIZE` is a thing I've been thinking I should just change it to

```
#if defined(SIMDE_NATURAL_VECTOR_SIZE)
  #define SIMDE_ASSUME_VECTORIZATION
#endif
```

Or maybe just `s/SIMDE_ASSUME_VECTORIZATION/SIMDE_NATURAL_VECTOR_SIZE/`.  IIRC that was actually one of the original use cases for `SIMDE_NATURAL_VECTOR_SIZE`, but I forgot about it when I merged the patch.

> but how would you have used SIMDE_PREFER_VECTOR_SIZE instead?

Just

```c
#elif SIMDE_PREFER_VECTOR_SIZE(128)
  return simde_vcgt_f32(a, simde_vdup_n_f32(SIMDE_FLOAT32_C(0.0)));
```

But I really don't think it adds anything here.  We don't want to restrict it to vector lengths <= 128, we want to always fall back.  That's what `SIMDE_ASSUME_VECTORIZATION` is for.

It seems like what we really want to do here is something like

```c
  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
    return vcgtz_f32(a);
  #elif defined(SIMDE_VECTOR_SUBSCRIPT_SCALAR)
    simde_float32x2_private a_ = simde_float32x2_to_private(a);
    simde_uint32x2_private r_;

    r_.values = HEDLEY_REINTERPRET_CAST(__typeof__(r_.values), a_.values > SIMDE_FLOAT32_C(0.0));

    return simde_uint64x1_from_private(r_);
  #elif defined(SIMDE_ASSUME_VECTORIZATION)
    return simde_vcgt_f32(a, simde_vdup_n_f32(SIMDE_FLOAT32_C(0.0)));
  #else
    simde_float32x2_private a_ = simde_float32x2_to_private(a);
    simde_uint32x2_private r_;

    SIMDE_VECTORIZE
    for (size_t i = 0 ; i < (sizeof(r_.values) / sizeof(r_.values[0])) ; i++) {
      r_.values[i] = (a_.values[i] > SIMDE_FLOAT32_C(0.0)) ? UINT32_MAX : 0;
    }

    return simde_uint32x2_from_private(r_);
  #endif
```

I'm a bit groggy right now, but I *think* we can safely change the vector extension part to just

```c
  #elif defined(SIMDE_VECTOR_SUBSCRIPT_SCALAR)
    HEDLEY_REINTERPRET_CAST(simde_uint32x2_t, a > SIMDE_FLOAT32_C(0.0));
```

which would make that a *lot* more palatable.

Reply:
> I'm a bit groggy right now, but I think we can safely change the vector extension part to just

I just thought about this a bit more; it wouldn't work reliably.  WASM is a counte-example :(

Maybe we should add functions like simde_float32x2_to_f32 if SIMDE_VECTOR_SUBSCRIPT is defined?  It would still be a bit ugly, but much better than what I wrote above…